staging/src/k8s.io/apiserver/pkg/server/connectivity_service.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/apiserver/pkg/apis/apiserver"
+	"k8s.io/klog"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+var directDialer utilnet.DialFunc
```

Comment:
We try to avoid globals.  If this isn't for pooling, can you instead make this a function to return a new instance and assign to the function so that it's still mockable, but we don't have the global?

```go
func defaultDirectDialer() utilnet.DialFunc{
    transport, ok := http.DefaultTransport.(*http.Transport)
    if ok {
        return transport.DialContext
    }
    panic("coder error")
}

var directDialerFn = defaultDirectDialer
```

Reply:
inlining as discussed.